### PR TITLE
Add Testtool for FIMDB features

### LINKED ISSUE #10226: Develop a testing tool for the FIM DBSync
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| [9103-replace-fim-db-dbsync](https://github.com/wazuh/wazuh/tree/9103-replace-fim-db-dbsync)  | FIMDB | Manager/Agent | Packages/Sources | All |

## Description

During the implementation of the new fim DB module, using the DBSync library, we have considered developing a test tool that allows us to check that the DB is working correctly, verifying that the events correctly generate the necessary entries in the DB, and that the synchronization with the manager occurs correctly.
This is a way to quickly check all the changes carried out during the development of DBSync.

### Possible capabilities
After several analyzed proposals and a lot of research for possible designs we have come up with some possibilities for the testtool:

- Mainly we want it to be a black box that is in charge of receiving an input, process what is needed from the new FIM DB module, and return a variable output that gives us to understand that everything has worked correctly.
For this, we can use a json file, with which we can configure the necessary input, passing it the actions that we want to check and even the specific data of the files with which to make the test. Then the tool would be in charge of processing and using the functions that are needed, and give us an output of the state of the database that allows us to verify that everything is correct.

- Another possibility could be to use a tool that directly receives as parameter the name of a function of the FIM DB module that we are interested in testing, as well as the necessary information of files/registries to work with.
Here we have the advantage of not needing a very complex logic that converts the input in different checks.

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|Closes #10226 |

## Description
During the implementation of the new fim DB module, using the DBSync library, we have considered developing a test tool that allows us to check that the DB is working correctly, verifying that the events correctly generate the necessary entries in the DB, and that the synchronization with the manager occurs correctly.
This is a way to quickly check all the changes carried out during the development of DBSync.

### CODE REVIEW

File: src/syscheckd/src/db/testtool/factoryAction.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+
+#ifndef _FACTORY_ACTION_H
+#define _FACTORY_ACTION_H
+#include <iostream>
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/testtool/factoryAction.h:14]: This header can be deleted

File: src/syscheckd/src/db/testtool/main.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+
+#include <fstream>
+#include <ostream>
+#include <stdio.h>
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/testtool/main.cpp:14]: This header can be deleted

File: src/syscheckd/src/db/testtool/main.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                    testContext->currentId = idx;
+                    const std::string inputFile{ actions[idx] };
+                    std::ifstream actionsIdxFile{ inputFile };
+                    std::cout << "Processing file: " << inputFile << std::endl;
+                    const auto& jsonAction { nlohmann::json::parse(actionsIdxFile) };
+                    auto action { FactoryAction::create(jsonAction.at("action").get<std::string>()) };
+                    action->execute(testContext, jsonAction.at("body"));
+                }
+
+                //DB::stopIntegrity();
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/testtool/main.cpp:104]: Here we can add a call to "DB::teardown()" in order to close correctly the database and synchronization

Comment (FrancoRivero) [inline src/syscheckd/src/db/testtool/main.cpp:104]: Here we can add a call to "DB::instance().teardown()" in order to close correctly the database and synchronization

File: src/syscheckd/src/db/testtool/main.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+ * Foundation.
+ */
+
+#include <fstream>
+#include <ostream>
+#include <stdio.h>
+#include <memory>
+#include <json.hpp>
+#include "db.hpp"
+#include "fimDB.hpp"
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/testtool/main.cpp:18]: This header can be deleted

**Review by FrancoRivero — APPROVED** [review]

**Review by antoniomanuelfr — APPROVED** [review]

**Review by jotacarma90 — APPROVED** [review]
LGTM

**Review by chemamartinez — APPROVED** [review]

File: src/syscheckd/src/db/testtool/action.h [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    {
+        auto retVal = false;
+        int count = 0;
+        try
+        {
+            const auto filterType
+            {
+                static_cast<COUNT_SELECT_TYPE>(value.at("filter_type").get<int32_t>())
+            };
+            count = DB::instance().countFiles(filterType);
```
Comment (FrancoRivero) [inline src/syscheckd/src/db/testtool/action.h:106]: This has changed in 9103, and the way to call it is : `count = DB::instance().countEntries(FIMDB_FILE_TABLE_NAME, COUNT_SELECT_TYPE::COUNT_ALL);`
After this change the filterType variable will be useless

**Review by FrancoRivero — CHANGES_REQUESTED** [review]

**Review by FrancoRivero — APPROVED** [review]

**Pull request merged**